shrinkpdf.sh

```diff
@@ -61,6 +61,25 @@ shrink ()
 	  "$1"
 }
 
+get_pdf_version() {
+  # $1 is the input file. The PDF version is contained in the
+  # first 1024 bytes and will be extracted from the PDF file.
+  version=$(cut -b "-1024" "$1" | sed -n '/%PDF-[0-9]\.[0-9]/{s/.*%PDF-/%PDF-/;p;q}' | cut -c "-8")
```

Comment:
The error message with `cut` (https://github.com/aklomp/shrinkpdf/actions/runs/3275070529/jobs/5392222355#step:5:7) comes from the first `cut` invocation in this line, I suppose. The `sed` command closes its incoming pipe after finding the string `%PDF-x.y`. `cut` obviously does not handle it gracefully…

Reply:
The whole expression looks a bit overcomplicated to me. What we are trying to accomplish is parse the PDF signature, which has a fixed location near the start of the file. There might be a way to get `sed` to directly return the first match in the first "line".

I also have the feeling that extraction and formatting could be handled nicely by an AWK one-liner.

Reply:
Let's say: I am not proud of it. I read, that the header can be anywhere within the first 1024 bytes of the file according to Adobes documents (see *Acrobat viewers require only that the header appear somewhere within the first 1024 bytes of the file* https://bugs.ghostscript.com/show_bug.cgi?id=687125 for example). I did not find the actual documentation by Adobe, I guess it is not available freely.

By prepending the `cut` command I wanted to ensure, that only a fixed amount of the file needs to be read by the function. But yeah, a AWK script could also do the trick and be easier on the eyes…